tests/array_list_with_map_test.c

```diff
@@ -0,0 +1,186 @@
+
+static int s_array_list_with_map_get_random_fn(struct aws_allocator *allocator, void *ctx) {
+    (void)ctx;
+    AWS_STATIC_STRING_FROM_LITERAL(foo, "foo");
+
+    struct aws_array_list_with_map list_with_map;
+    /* With only 1 initial element. */
+    ASSERT_SUCCESS(aws_array_list_with_map_init(
+        &list_with_map, allocator, s_hash_string_ptr, s_hash_string_ptr_eq, NULL, sizeof(struct aws_string *), 1));
+    struct aws_string *left_element = NULL;
+    /* Fail to get any, when there is nothing in it. */
+    ASSERT_FAILS(aws_array_list_with_map_get_random(&list_with_map, &left_element));
+    ASSERT_SUCCESS(aws_array_list_with_map_insert(&list_with_map, &foo));
+
+    /* Check the size */
+    ASSERT_UINT_EQUALS(aws_array_list_with_map_length(&list_with_map), 1);
+    ASSERT_SUCCESS(aws_array_list_with_map_get_random(&list_with_map, &left_element));
+    ASSERT_TRUE(aws_string_eq(left_element, foo));
+
+    aws_array_list_with_map_clean_up(&list_with_map);
+    return AWS_OP_SUCCESS;
+}
+
+AWS_TEST_CASE(array_list_with_map_get_random_test, s_array_list_with_map_get_random_fn)
+
+static int s_array_list_with_map_exist_fn(struct aws_allocator *allocator, void *ctx) {
+    (void)ctx;
+    AWS_STATIC_STRING_FROM_LITERAL(foo, "foo");
+    AWS_STATIC_STRING_FROM_LITERAL(bar, "bar");
+
+    struct aws_array_list_with_map list_with_map;
+    ASSERT_SUCCESS(aws_array_list_with_map_init(
+        &list_with_map, allocator, s_hash_string_ptr, s_hash_string_ptr_eq, NULL, sizeof(struct aws_string *), 1));
+    ASSERT_SUCCESS(aws_array_list_with_map_insert(&list_with_map, &foo));
+
+    ASSERT_TRUE(aws_array_list_with_map_exist(&list_with_map, &foo));
+    ASSERT_FALSE(aws_array_list_with_map_exist(&list_with_map, &bar));
+
+    aws_array_list_with_map_clean_up(&list_with_map);
+    return AWS_OP_SUCCESS;
+}
+
+AWS_TEST_CASE(array_list_with_map_exist_test, s_array_list_with_map_exist_fn)
+
+static int s_array_list_with_map_remove_fn(struct aws_allocator *allocator, void *ctx) {
+    (void)ctx;
+    (void)ctx;
+    AWS_STATIC_STRING_FROM_LITERAL(foo, "foo");
+    AWS_STATIC_STRING_FROM_LITERAL(bar, "bar");
+    AWS_STATIC_STRING_FROM_LITERAL(foobar, "foobar");
+
+    struct aws_array_list_with_map list_with_map;
+    /* With only 1 initial element. */
+    ASSERT_SUCCESS(aws_array_list_with_map_init(
+        &list_with_map, allocator, s_hash_string_ptr, s_hash_string_ptr_eq, NULL, sizeof(struct aws_string *), 1));
+    ASSERT_SUCCESS(aws_array_list_with_map_insert(&list_with_map, &bar));
+    ASSERT_SUCCESS(aws_array_list_with_map_insert(&list_with_map, &foobar));
+    ASSERT_SUCCESS(aws_array_list_with_map_insert(&list_with_map, &foo));
+
+    ASSERT_SUCCESS(aws_array_list_with_map_remove(&list_with_map, &foo));
+    /* Check the size */
+    ASSERT_UINT_EQUALS(aws_array_list_with_map_length(&list_with_map), 2);
+
+    /* Should be fine to remove something not existing anymore? */
```

Comment:
Should succeed and do nothing:
https://github.com/awslabs/aws-c-common/blob/d9031cf4f02fe4d8c24a53bb6c8df67ce8be9f0c/include/aws/common/hash_table.h#L279